### PR TITLE
***WIP*** Validation rebase ***WIP***

### PR DESCRIPTION
**Pull request by yanaise**
Validate cache consistency 

---

---
_Release Notes_: 
Should avoid cache consistency problems, would not be noticed by clients

### CODE REVIEW

File: packages/workspace/src/workspace/nacl_files/elements_cache.ts [hunk 1]
```diff
@@ -127,3 +135,161 @@ export const getAfterElements = async ({
...
+  }> => {
+  const elementsToMerge: (Element | undefined)[] = []
+  const deletedIds = new Set<string>()
+  await Promise.all(srcChanges.changes.map(async change => {
+    const element = getChangeElement(change)
+    const id = element.elemID
+    if (isAdditionOrModificationChange(change)) {
+      elementsToMerge.push(element)
+    } else if (isRemovalChange(change)) {
+      deletedIds.add(id.getFullName())
```
Comment (roironn) [inline packages/workspace/src/workspace/nacl_files/elements_cache.ts:172]: This means that we will add an id to deleted IDs even if it has an unmodified fragment... Is this the desired behaviour? If so - maybe we need another name for the var :)

File: packages/workspace/src/workspace/nacl_files/elements_cache.ts [hunk 2]
```diff
@@ -127,3 +135,161 @@ export const getAfterElements = async ({
...
+  deletedIds: Set<string>): Promise<{
+    mergedChanges: Change[]
+    mergeErrors: RemoteMapEntry<MergeError[], string>[]
+  }> => {
+  const mergeErrors: RemoteMapEntry<MergeError[], string>[] = []
+  const sieve = new Set<string>()
+  const mergedChanges = (await awu(newMergedElementsResult.merged.values()).map(async element => {
+    const id = element.elemID
+    const fullname = id.getFullName()
+    deletedIds.delete(fullname)
```
Comment (roironn) [inline packages/workspace/src/workspace/nacl_files/elements_cache.ts:196]: ?

File: packages/workspace/src/workspace/nacl_files/elements_cache.ts [hunk 3]
```diff
@@ -127,3 +135,161 @@ export const getAfterElements = async ({
...
+export const mergeChanges = async ({
+  src1Changes,
+  src2Changes,
+  src1,
+  src2,
+  currentElements,
+  cachePreChangeHash,
+  mergeFunc,
+}: {
+  src1Changes: ChangeSet<Change<Element>>
```
Comment (roironn) [inline packages/workspace/src/workspace/nacl_files/elements_cache.ts:245]: Should probably define a type here for politeness :)

File: packages/workspace/src/workspace/nacl_files/elements_cache.ts [hunk 4]
```diff
@@ -127,3 +135,161 @@ export const getAfterElements = async ({
...
+  src2: ReadOnlyElementsSource
+  currentElements: ElementsSource
+  cachePreChangeHash: string | undefined
+  mergeFunc: (elements: AsyncIterable<Element>) => Promise<MergeResult>
+}): Promise<{
+  mergedChanges: ChangeSet<Change<Element>>
+  mergeErrors: AsyncIterable<RemoteMapEntry<MergeError[], string>>
+}> => {
+  const src1ChangeIDs = new Set<string>()
+  src1Changes.changes.forEach(change => src1ChangeIDs
```
Comment (roironn) [inline packages/workspace/src/workspace/nacl_files/elements_cache.ts:257]: better use map and pass the array to the set?

File: packages/workspace/src/workspace/nacl_files/elements_cache.ts [hunk 5]
```diff
@@ -127,3 +135,161 @@ export const getAfterElements = async ({
...
+    .add(getChangeElement(change).elemID.getFullName()))
+  const src2ChangeIDs = new Set<string>()
+  src2Changes.changes.forEach(change => src2ChangeIDs
```
Comment (roironn) [inline packages/workspace/src/workspace/nacl_files/elements_cache.ts:260]: same as above

File: packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts [hunk 6]
```diff
@@ -172,9 +170,20 @@ const buildMultiEnvSource = (
...
+      mergedChanges: changeResult.mergedChanges,
+      mergeErrors: changeResult.mergeErrors,
+      currentElements: current.elements,
+      currentErrors: current.mergeErrors,
+    })
     return {
       state: current,
-      changes: mergeChanges,
+      changes: {
+        changes: changeResult.mergedChanges.changes as Change<ChangeDataType>[],
```
Comment (roironn) [inline packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts:182]: 1. Why do we need the cast?
2. better use {... changeResult.mergedChanges, changes: ... } if we must have the cast.

File: packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts [hunk 7]
```diff
@@ -222,18 +231,22 @@ const buildMultiEnvSource = (
...
   Promise<EnvsChanges> => {
     const secondaryChanges = routedChanges.secondarySources || {}
-    return resolveValues({
-      [primarySourceName]: primarySource().updateNaclFiles(routedChanges.primarySource || []),
-      [commonSourceName]: commonSource().updateNaclFiles(routedChanges.commonSource || []),
-      ..._.mapValues(secondaryChanges, (changes, srcName) =>
-        secondarySources()[srcName].updateNaclFiles(changes)),
-    })
+    return {
+      ...(await resolveValues({
```
Comment (roironn) [inline packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts:235]: The two awaits here will create extra requests in the SaaS UI, so better change it Promise all with one await here.

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 8]
```diff
@@ -237,8 +239,10 @@ const buildNaclFilesState = async ({
   newNaclFiles: ParsedNaclFile[]
   currentState: NaclFilesState
 }): Promise<buildNaclFilesStateResult> => {
+  const preChangeHash = await currentState.parsedNaclFiles.getHash()
   log.debug('building elements indices for %d NaCl files', newNaclFiles.length)
   const newParsed = _.keyBy(newNaclFiles, parsed => parsed.filename)
+  let cacheValid = true
```
Comment (roironn) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:245]: Based on the usage below (inside the current file check validaity I think we need to use a more expressive name (sourceCacheValid maybe? globalCacheValid? Something like that...)

File: packages/workspace/src/workspace/workspace.ts [hunk 9]
```diff
@@ -325,11 +324,15 @@ export const loadWorkspace = async (
...
     }
-
-    const mergeData = await getAfterElements({
+    const changeResult = await mergeChanges({
       src1Changes: workspaceChanges,
       src1: await naclFilesSource.getElementsSource(),
-      src2Changes: await completeStateOnlyChanges(stateOnlyChanges),
+      src2Changes: {
+        changes: await completeStateOnlyChanges(stateOnlyChanges.changes),
+        cacheValid: stateOnlyChanges.cacheValid,
```
Comment (roironn) [inline packages/workspace/src/workspace/workspace.ts:332]: unpack would be more readable IMP

File: packages/workspace/src/workspace/workspace.ts [hunk 10]
```diff
@@ -436,20 +444,26 @@ export const loadWorkspace = async (
...
-      .updateNaclFiles(visibleChanges, mode)
+    const workspaceChanges = await ((await getLoadedNaclFilesSource())
+      .updateNaclFiles(visibleChanges, mode))
     const stateOnlyChanges = await getStateOnlyChanges(hiddenChanges)
-    workspaceState = buildWorkspaceState({ workspaceChanges, stateOnlyChanges, validate })
-    return (workspaceChanges.length + stateOnlyChanges.length)
+    workspaceState = buildWorkspaceState({ workspaceChanges,
+      stateOnlyChanges: {
+        changes: stateOnlyChanges,
+        cacheValid: true,
```
Comment (roironn) [inline packages/workspace/src/workspace/workspace.ts:453]: why is this hardcoded? Can't the state cache be invalid?

File: packages/workspace/src/workspace/nacl_files/elements_cache.ts [hunk 11]
```diff
@@ -127,3 +135,166 @@ export const getAfterElements = async ({
...
+  currentErrors: RemoteMap<SaltoError[]>
+}): Promise<void> => {
+  await currentErrors.setAll(mergeErrors)
+  await currentElements.setAll(awu(mergedChanges.changes).filter(isAdditionOrModificationChange)
+    .map(change => getChangeElement(change)))
+  await currentElements.deleteAll(awu(mergedChanges.changes).filter(isRemovalChange)
+    .map(change => getChangeElement(change).elemID))
+}
+
+const readSourceChangesAndAddUnmodifiedFragments = async (
```
[the remark this answers is not in the document]
Comment (yanaise) [reply packages/workspace/src/workspace/nacl_files/elements_cache.ts:157]: nice idea, but it wouldn't work. if you take a look at the original function, it sometimes pushes one element, sometimes zero, and sometimes two. Sometimes we need to push both an unmodified fragment and an element, and that can't happen with a simple map. I can do it with a flatmap, where sometimes I return two elements if you think that is more readable or efficient, so let me know if you think that would be the right way.

File: packages/workspace/src/workspace/workspace.ts (hunk 10 above)
Comment (yanaise) [reply packages/workspace/src/workspace/workspace.ts:453]: At first glance, I couldn't find a way for it to be invalid, but I'd like to zoom about it, maybe there's something.

**Pull request closed without merging**